### PR TITLE
Add :print-length and :print-level options

### PR DESCRIPTION
**Pull request by cichli**
Hi, first of all, thanks for the great (and _fast_!) library :-).

Both options are analogous to (and default to) `*print-length*` and `*print-level*` respectively.

I also took a stab at improving the benchmark, using `test.check` to generate sample data using `any-printable`. Additionally, I've switched to using a stubbed `Writer` implementation instead of `with-out-str` to try and minimise any overhead.

Here are results for the new benchmark, both [before](https://gist.github.com/cichli/2b99d646f72f0ddbb5e8) and [after](https://gist.github.com/cichli/01e5049ef3dd764555f2) this change is applied - the impact is reasonably negligible for normal usage (i.e. when `:print-length` / `:print-level` are not set). Please do try and verify these results locally!

Notes on the implementation:

* `:print-length` is implemented by using a `take` transducer to limit the number of lines returned by the relevant `visit-*` methods, iff `:print-length` is set.
* The string ` ...` is appended to the collection being printed if its length exceeds `:print-length`.

* `:print-level` is implemented similarly to in Clojure itself - each time we call one of the relevant `visit-*` methods, we decrement its value, iff `:print-level` is set. This is done by using a new instance of `EdnPrinter` rather than the dynamic binding approach used in Clojure.
* The string `#` is printed instead of the contents of the collection if the current `:print-level` is negative.

Motivation for this change: we now use `fipp` as the default pretty-printer in CIDER, but would like a way to prevent accidental printing of infinitely long or recursive structures :-). Let me know if there's anything I've missed or you think I should change.

### CODE REVIEW

**Comment by brandonbloom** [comment]
Hi Michael,

It's very exciting to hear that Fipp is now the default in Cider! Thanks for taking on this work.

I might not have a chance for a few days to pull down the code and play with it. I'll take a super quick look at the code now and make some comments. While it's OK by me if the engine code gets a little hairy, I'd like to do my best to keep the pretty-document printer code as clean as possible, as it's a good reference for people writing printers for non-edn.

Cheers,
Brandon

File: src/fipp/edn.cljc [hunk 1]
```diff
@@ -38,18 +55,41 @@
   (visit-seq [this x]
     (if-let [pretty (symbols (first x))]
       (pretty this x)
-      [:group "(" [:align (interpose :line (map #(visit this %) x))] ")"]))
+      (let [printer (cond-> this
```
Comment (brandonbloom) [inline src/fipp/edn.cljc]: This pattern repeats for seqs, vectors, maps, and sets. Could it be abstracted in to a `print-coll` function?

File: test/fipp/benchmark.clj [hunk 2]
```diff
@@ -1,26 +1,42 @@
...
+  (proxy [Writer] []
+    (write [_])
+    (flush [])
+    (close [])))
+
+(defn samples []
+  (let [^File file (io/file ".benchmark-samples")]
+    (when-not (.exists file)
+      (spit file (with-out-str
+                   (fipp.edn/pprint (gen/sample gen/any-printable 1000)))))
```
Comment (brandonbloom) [inline test/fipp/benchmark.clj]: I haven't actually used test.check in anger yet. I'd love to keep the randomness out of the _benchmarks_ (it's fine/welcome in correctness tests). Is it possible to force the seed? Even if yes, why is randomly generated printable data superior for testing than a realistic code file?

File: test/fipp/benchmark.clj [hunk 3]
```diff
@@ -1,26 +1,42 @@
...
 (defn -main []
-  (doseq [f benched-fns
-          [k v] samples]
-    (println "Testing " f " on " k)
-    (criterium.core/with-progress-reporting
-      (bench (with-out-str (f v))))))
+  (let [samples (samples)]
+    (doseq [f benched-fns]
+      (println "Benchmarking " f)
+      (time
```
Comment (brandonbloom) [inline test/fipp/benchmark.clj]: Doesn't bench already print time?

**Comment by brandonbloom** [comment]
Looking at this code more carefully now... additional comments follow.

File: src/fipp/edn.cljc (hunk 1 above)
Comment (brandonbloom) [inline src/fipp/edn.cljc]: I've made this change locally, will publish soon.

File: test/fipp/benchmark.clj (hunk 2 above)
Comment (brandonbloom) [inline test/fipp/benchmark.clj]: I didn't read this carefully enough, I see you're saving the samples file. Makes sense, but I still need to give more thought to whether we want to test random stuff or not.

File: src/fipp/edn.cljc [hunk 4]
```diff
@@ -5,7 +5,24 @@
...
+(defn- ->lines [{:keys [print-level] :as printer} xform x]
+  (if (and print-level (neg? print-level))
+    "#"
+    (sequence xform x)))
+
+(defn- ->group [{:keys [print-length] :as printer} left-separator lines right-separator]
+  [:group left-separator
+          [:align lines]
+          (when (and print-length (<= print-length (count lines)))
+            " ...")
```
Comment (brandonbloom) [inline src/fipp/edn.cljc]: Actually want `[:span :line "..."]` here, or more generally `[:span sep "..."]`. This change plus corresponding tests forthcoming.

**Comment by brandonbloom** [comment]
Removed the benchmarking commit, squashed the remaining commits, cleaned up the code a bit, fixed some edge cases, updated the docs.

@cichli Please try the master branch and make sure it still works as you'd expect. If it does, I'll cut a release.

**Comment by cichli** [comment]
Hi Brandon,

Thanks a lot for taking the time to look at the PR - and then to subsequently clean it up and merge it! `pretty-coll` is indeed a much cleaner impl. I've checked out the changes locally and they work exactly as I'd expect. Looking forward to the release!

I'll add my part to the benchmarking discussion on the other PR.

Thanks,

Michael

**Comment by brandonbloom** [comment]
OK released version 0.6.3 - Thanks again!
